### PR TITLE
Limit and Offset for pagination

### PR DESCRIPTION
**Pull request by Henrod**
## Pagination

Pagination on GET /offers

 - On GET /offers, use parameters _limit_ to tell the page size, and _offset_ to tell which page to return. 
- If limit is not informed, the default is 50
- If offset is not informed, the default is 0
- The GET /offers response now is a JSON with the following keys:
  - offers: list of offers
  - page: total number of pages given the limit

### CODE REVIEW

File: api/offer.go [hunk 1]
```diff
@@ -177,10 +183,28 @@ func (g *OfferHandler) list(w http.ResponseWriter, r *http.Request) {
 		return
 	}
 
-	var err error
+	var limit uint64
+	if limitStr == "" {
+		limit = 50
+	} else if limit, err = strconv.ParseUint(limitStr, 10, 64); err != nil {
```
Comment (cscatolini) [inline api/offer.go]: I'd separate this in more than one statement:

```
var limit uint64
if limitStr == "" {
  limit = 50
} else {
  var err error
  limit, err = strconv.ParseUint(limitStr, 10, 64)
  if err != nil {
    logger.WithError(err).Error("List game offers failed.")
    g.App.HandleError(w, http.StatusBadRequest, "The limit parameter must be an uint.", err)
    return
  }
}
```

File: api/offer.go [hunk 2]
```diff
@@ -177,10 +183,28 @@ func (g *OfferHandler) list(w http.ResponseWriter, r *http.Request) {
...
+		logger.WithError(err).Error("List game offers failed.")
+		g.App.HandleError(w, http.StatusBadRequest, "The limit parameter must be an uint.", err)
+		return
+	}
+
+	var offset uint64
+	if offsetStr == "" {
+		offset = 0
+	} else if offset, err = strconv.ParseUint(offsetStr, 10, 64); err != nil {
```
Comment (cscatolini) [inline api/offer.go]: same here

File: api/offer.go [hunk 3]
```diff
@@ -177,10 +183,28 @@ func (g *OfferHandler) list(w http.ResponseWriter, r *http.Request) {
 		return
 	}
 
-	var err error
+	var limit uint64
+	if limitStr == "" {
+		limit = 50
```
Comment (cscatolini) [inline api/offer.go]: we'd better retrieve this from the config

File: api/offer.go [hunk 4]
```diff
@@ -191,10 +215,16 @@ func (g *OfferHandler) list(w http.ResponseWriter, r *http.Request) {
...
+		"offers": offers,
+		"pages":  pages,
+	}
+
+	bts, _ := json.Marshal(responseObj)
+	status := http.StatusOK
 	if len(offers) == 0 {
-		Write(w, http.StatusOK, "[]")
-		return
+		status = http.StatusNoContent
```
Comment (cscatolini) [inline api/offer.go]: if `len(offers) == 0` we should still return 200 and an object like this:

```
{
  "offers": [],
  "pages": 0
}
```

File: models/offer.go [hunk 5]
```diff
@@ -144,16 +144,49 @@ func GetEnabledOffers(db runner.Connection, gameID string, offersCache *cache.Ca
...
+			Where("game_id = $1", gameID).
+			QueryScalar(&numberOffers)
+	})
+	if err != nil {
+		return offers, 0, err
+	}
+
+	var pages int
+	if limit != 0 {
+		pages = numberOffers / int(limit)
```
Comment (cscatolini) [inline models/offer.go]: instead of doing this + the if below isn't it simpler to divide using floats and round it up (ceil)?

Comment (andrehp) [inline models/offer.go]: I don't really like using floats where they aren't strictly necessary, so in this case I prefer the integer division approach.

Comment (cscatolini) [inline models/offer.go]: @Henrod you can go with this approach then

**Pull request merged**